draft/XXXX-robust-api/README.md

```diff
@@ -0,0 +1,146 @@
+
+```
+{
+  "uploadPartTags": [
+    <ETag returned from the first signed url>,
+    <ETag returned from the second signed url>,
+    ...
+  ]
+}
+```
+
+**NOTE:** Similar to the `uploadPartUrls` array mentioned in the Upload creation step, the ordering of the `uploadPartTags` array is significant.
+
+At this point, the sever downloads the reconstructed file from S3, and performs data processing, table/graph creation, etc. with the appropriate processor for that data type. This process doesn't technically need to be done asynchronously, as if the data is small enough it's feasible that this could be completed within the time frame of a single request. However, in the interest of scalability and robustness, this HTTP request should initiate an asynchronous task to process the data.
+
+## Additional Changes
+
+There are changes in addition to the endpoints mentioned above, which may not be crucial to the client, but should be detailed.
+
+### Upload Collection
+
+Each workspace should contain its own upload collection, which will be used to store each upload document shown above. This will need to be accompanied by a new `Upload` ORM, which will facilitate operations on uploads.
+
+### Additional Upload Endpoints
+
+`GET /api/workspaces/{workspace}/upload/{id}`
+
+This endpoint returns the Upload document with the corresponding ID.
+
+## Open Questions
+
+### Changes to data processing procedure
+
+The notion of having various file type "uploaders" will be removed. Instead, these will become data processors, which will take raw uploaded data (from S3), and run validation, table/graph creation, data insertion, etc. When downloading the finalized upload from S3, this processing should be done in a streaming manner, in order to limit the risk of running out of memory in the server. This will involve updating each data processor to handle streamed data, instead of the entire file. This could be done in the following manner:
+
+1. Read in data until a fully formed object can be parsed (a table row, for example).
+2. Run validation on this object
+3. If validation is passed, insert this object into the target location. If validation fails, save the validation error, and either continue processing more data, or halt data processing.
+4. Repeat steps 1-3 until all data has been processed
+5. If there are errors resulting from this processing, delete the partially formed tables/graphs, and return the validation errors to the user.
+
+These errors would likely be stored in the upload model (yet to be outlined), however, it may make sense to introduce another model that would track this status more specifically. Feedback is welcome here.
+
+
+### Tracking upload progress
+
+The current proposal requires that the client upload all parts, and then returns the ETags of these parts all at once when finalizing the upload. However, this poses a couple of possible pitfalls:
```

Comment:
Another is, what if one part fails out of 1000? That would introduce a lot of error checking logic on the client side so I don't think your first concept is reliable enough. This second idea, to use the `finalizePart` endpoint seems much more reliable

Reply:
I'm a bit confused, what error checking logic would need to be introduced? If any of the upload parts fails, I'd think the client would just declare the upload as failed. Regardless, I do agree the second approach is probably more flexible.

Reply:
I meant if the client sends a finalize without all the parts being done (or one failing). Then the server would error when it tries to read the data in. The solution would be to set the client logic to not send the finalize, but that seems less robust than the server keeping track and rejecting a finalize until all the parts have settled.

Error checking was a bit of a misnomer, because it has a specific meaning in this context. I meant just checking for a missed upload part

Reply:
> I meant if the client sends a finalize without all the parts being done (or one failing). Then the server would error when it tries to read the data in. The solution would be to set the client logic to not send the finalize, but that seems less robust than the server keeping track and rejecting a finalize until all the parts have settled.

My proposal was actually _not_ to have the client send the finalize request, but rather, have the server send the request. All the client would need to do is to send the appropriate `Etags` on upload part completion. I believe this split between access to upload parts and access to the entire upload object can be enforced through permissions on the S3 object.

Reply:
I'm not sure if I understand the problem: for the client to finalize, it has to have all of the response ETags, which means all the part uploads succeeded. Right?

Reply:
Yes. From the multipart upload docs:

> Your complete multipart upload request must include the upload ID and a list of both part numbers and corresponding ETag values. Amazon S3 response includes an ETag that uniquely identifies the combined object data. 

Which we would do through an Python S3 driver (`boto3`).

Reply:
> My proposal was actually not to have the client send the finalize request, but rather, have the server send the request. All the client would need to do is to send the appropriate Etags on upload part completion.

Ah, got it. Thanks for the clarification